### PR TITLE
chore: Remove the Metrics name dedup

### PR DESCRIPTION
**Pull request by sarabala1979**

### CODE REVIEW

**Comment by sarabala1979** [comment]
I created the [issue](https://github.com/numaproj/numaflow-sinks/issues/23) to add test case for the Prometheus push. We need to find the test library to test Prometheus abstraction. we need to refactor the code

File: prometheus-pusher/main.go [hunk 1]
```diff
@@ -55,32 +55,30 @@ func (c *myCollector) Collect(ch chan<- prometheus.Metric) {
 }
 
 func (p *prometheusSink) push(msgPayloads []Payload) error {
-	keys := make(map[string]bool)
-
 	for _, payload := range msgPayloads {
-		if _, ok := keys[payload.Name]; !ok {
```
Comment (sarabala1979) [inline prometheus-pusher/main.go:61]: name level dedub was removed

File: prometheus-pusher/main.go [hunk 2]
```diff
@@ -55,32 +55,30 @@ func (c *myCollector) Collect(ch chan<- prometheus.Metric) {
...
-				appName := payload.Labels["app"]
-				p.metrics.IncreaseAnomalyGenerated(payload.Namespace, appName, payload.Name)
-				keys[payload.Name] = true
-			}
-			err = pusher.Push()
-			if err != nil {
-				return err
-			}
-			p.metrics.IncreaseTotalSuccess()
+		p.logger.Debug("Pushing Payload ", zap.Any("payload", payload))
```
Comment (whynowy) [inline prometheus-pusher/main.go:59]: `Debugw`

File: prometheus-pusher/main.go [hunk 3]
```diff
@@ -55,32 +55,30 @@ func (c *myCollector) Collect(ch chan<- prometheus.Metric) {
...
+		case "Gauge":
+			p.logger.Debugf("Creating Collector %s", payload.Name)
+			pusher = pusher.Collector(&myCollector{
+				metric:     prometheus.NewDesc(payload.Name, "", nil, payload.Labels),
+				metricType: prometheus.GaugeValue,
+				value:      payload.Value,
+			})
+			appName := payload.Labels["app"]
+			p.metrics.IncreaseAnomalyGenerated(payload.Namespace, appName, payload.Name)
+		}
```
Comment (whynowy) [inline prometheus-pusher/main.go:74]: Logically, missing `default`?

File: prometheus-pusher/main.go [hunk 4]
```diff
@@ -55,32 +55,30 @@ func (c *myCollector) Collect(ch chan<- prometheus.Metric) {
...
-				p.metrics.IncreaseAnomalyGenerated(payload.Namespace, appName, payload.Name)
-				keys[payload.Name] = true
-			}
-			err = pusher.Push()
-			if err != nil {
-				return err
-			}
-			p.metrics.IncreaseTotalSuccess()
+		p.logger.Debug("Pushing Payload ", zap.Any("payload", payload))
+		pusher, err := p.createPusher(fmt.Sprintf("%s_%s_%s", payload.Namespace, payload.Subsystem, payload.Name))
```
Comment (qhuai) [inline prometheus-pusher/main.go:60]: Just to double check: do we need to create a new pusher for each payload item?

**Review by whynowy — APPROVED** [review]

File: prometheus-pusher/main.go [hunk 5]
```diff
@@ -55,32 +55,33 @@ func (c *myCollector) Collect(ch chan<- prometheus.Metric) {
...
+				metricType: prometheus.GaugeValue,
+				value:      payload.Value,
+			})
+			appName := payload.Labels["app"]
+			p.metrics.IncreaseAnomalyGenerated(payload.Namespace, appName, payload.Name)
+		default:
+			p.logger.Errorw("Unsupported Metrics Type", zap.Any("payload", payload))
+			return fmt.Errorf("unsupported Metrics Type")
+		}
+		err = pusher.Push()
```
Comment (qhuai) [inline prometheus-pusher/main.go:78]: Just to double check: if a pusher needs an explicit close or such call.

File: prometheus-pusher/main.go [hunk 6]
```diff
@@ -55,32 +55,33 @@ func (c *myCollector) Collect(ch chan<- prometheus.Metric) {
...
+		if err != nil {
+			p.logger.Errorw("Failed to push", zap.Any("payload", payload), zap.Error(err))
+			return err
 		}
+		p.logger.Infow("Successfully pushed", zap.Any("payload", payload))
```
Comment (qhuai) [inline prometheus-pusher/main.go:83]: would this infow() generate too many log lines in regular cases?

**Pull request merged**